nexus/db-queries/src/db/datastore/scim.rs

```diff
@@ -0,0 +1,157 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! [`DataStore`] methods related to SCIM
+
+use super::DataStore;
+use crate::authz;
+use crate::context::OpContext;
+use crate::db::model::ScimClientBearerToken;
+use async_bb8_diesel::AsyncRunQueryDsl;
+use chrono::Utc;
+use diesel::prelude::*;
+use nexus_db_errors::ErrorHandler;
+use nexus_db_errors::public_error_from_diesel;
+use nexus_db_lookup::LookupPath;
+use omicron_common::api::external::CreateResult;
+use omicron_common::api::external::DeleteResult;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupResult;
+use rand::{RngCore, SeedableRng, rngs::StdRng};
+use uuid::Uuid;
+
+// XXX this is the same as generate_session_token!
+fn generate_scim_client_bearer_token() -> String {
+    let mut rng = StdRng::from_os_rng();
+    let mut random_bytes: [u8; 20] = [0; 20];
+    rng.fill_bytes(&mut random_bytes);
+    hex::encode(random_bytes)
+}
+
+impl DataStore {
+    // SCIM tokens
+
+    pub async fn scim_idp_get_tokens(
+        &self,
+        opctx: &OpContext,
+        authz_silo: &authz::Silo,
+    ) -> ListResultVec<ScimClientBearerToken> {
+        opctx.authorize(authz::Action::ListChildren, authz_silo).await?;
```

Comment:
This seems to me too lenient of a check. Any silo viewer has `ListChildren` on the silo. Maybe that's fine considering the list of tokens doesn't actually tell you anything. But it's something to think about.

https://github.com/oxidecomputer/omicron/blob/ef318b326abd25f7d3fd7efce4363cc07043d937/nexus/db-queries/tests/output/authz-roles.out#L183-L195

The only way I know of to make this more restrictive is to create a "synthetic" authz resource representing the list of scim tokens for a given silo and then do the authz checks against that. Here's an example of a synthetic authz list resource with its own permissions distinct from silo `ListChildren`:

https://github.com/oxidecomputer/omicron/blob/ef318b326abd25f7d3fd7efce4363cc07043d937/nexus/auth/src/authz/api_resources.rs#L672-L695

Reply:
changed in 3b3b89190, have a look